pkg/pattern/tokenization/replacer.go

```diff
@@ -0,0 +1,859 @@
+package tokenization
+
+import (
+	"bytes"
+	"slices"
+)
+
+var (
+	placeholderNumber    = []byte("<NUM>")
+	placeholderHex       = []byte("<HEX>")
+	placeholderUUID      = []byte("<UUID>")
+	placeholderTimestamp = []byte("<TIMESTAMP>")
+	placeholderDuration  = []byte("<DURATION>")
+	placeholderBytesize  = []byte("<BYTESIZE>")
+	placeholderIP        = []byte("<IP>")
+)
+
+// Integer numbers after these words won't be replaced by `<NUM>`.
+var numPreservingKeys = [][]byte{[]byte("status"), []byte("status_code"), []byte("httpStatus")}
+
+const (
+	maxYear   = 2100 // anything above that is unlikely to be a timestamp...
+	minHexLen = 12   // 48 bits min for free-standing hex strings (e.g "0123456789ab") or 42 bits for "0xABCDEF1234" strings
+)
+
+var boundaryChars = [256]bool{}
+
+func init() {
+	for i := 0; i < 128; i++ { // for now, we don't consider non-ASCII chars as boundary
+		if i < '0' || (i > '9' && i < 'A') || (i > 'Z' && i < 'a') || i > 'z' {
+			boundaryChars[i] = true
```

Comment:
I am not certain if the slice lookups like `if boundaryChars[c] { /* do something */}` in the code below are actually faster than if we had a simple function like this:
```go
func isBoundary(c byte) {
	return c < '0' || (c > '9' && c < 'A') || (c > 'Z' && c < 'a') || (c > 'z' && c < 128)
}
```

I suspect the function will be faster because of how CPU caching works, but I haven't actually benchmarked it. I adopted this approach originally because I had a couple of exceptions here (e.g. I didn't consider `_` to be a boundary character initially).